lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,378 @@
+
+    """
+
+    def __init__(self, max_height_diff=35, nbhood_radius=1,
+                 max_lapse_rate=-3*DALR, min_lapse_rate=DALR):
+        """Initialise class.
+        Args:
+            max_height_diff (float):
+                Maximum allowable height difference between the central point
+                and points in the neighbourhood over which the lapse rate will
+                be calculated (metres).
+
+            nbhood_radius (int):
+                Radius of neighbourhood around each point. Note that the
+                neighbourhood will still be a square array.
+
+            max_lapse_rate (float):
+                Maximum lapse rate allowed.
+
+            min_lapse_rate (float):
+                Minimum lapse rate allowed.
+
+        Raises:
+            ValueError: If the nbhood_size is not an odd number.
+
+        """
+
+        self.max_height_diff = max_height_diff
+        self.max_lapse_rate = max_lapse_rate
+        self.min_lapse_rate = min_lapse_rate
+
+        # nbhood_size=3 corresponds to a 3x3 array centred on the
+        # central point.
+        self.nbhood_size = int((2*nbhood_radius) + 1)
+
+        # generic_filter extracts the neighbourhood and returns a 1D array.
+        # ind_central_point indicates where the central point would be on
+        # this array
+        self.nbhoodarray_size = self.nbhood_size**2
+        self.ind_central_point = int(self.nbhoodarray_size/2)
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        desc = ('<LapseRate>')
+        return desc
+
+    def _calc_lapse_rate(self, temperature, orography):
+        """Function to calculate the lapse rate.
+        This holds the function to determine the local lapse rate at a point by
+        calculating a least-squares fit to local temperature and altitude data
+        to find the local lapse rate.
+
+        Args:
+            temperature(1D np.array):
+            Contains the temperature values for the central point and its
+            neighbours.
+
+            orography(1D np.array):
+            Contains the height values for the central point and its
+            neighbours.
+
+        Returns:
+            gradient (float):
+            The gradient of the temperature/orography values. This
+            represents the lapse rate.
+
+        """
+
+        # If central point NaN then return blank value.
+        if np.isnan(temperature[self.ind_central_point]):
+            return DALR
+
+        # Remove points where there are NaN temperature values from both arrays
+        # before calculation.
+        y_data = temperature[~np.isnan(temperature)]
+        x_data = orography[~np.isnan(temperature)]
+
+        # Return DALR if standard deviation of both datasets = 0 (where all
+        # points are the same value).
+        if round(np.std(x_data), 6) == 0.0 and round(np.std(y_data), 6) == 0.0:
+            return DALR
```

Comment:
Might be more standard to use `np.isclose(np.std(x_data), 0)` for both float comparisons here.